cli/lib/kontena/cli/stacks/upgrade_command.rb

```diff
@@ -19,152 +22,197 @@ class UpgradeCommand < Kontena::Command
 
     option '--force', :flag, 'Force upgrade'
     option '--skip-dependencies', :flag, "Do not install any stack dependencies"
-    option '--dry-run', :flag, "Do not perform any uploading", hidden: true
+    option '--dry-run', :flag, "Simulate upgrade"
 
     requires_current_master
     requires_current_master_token
 
-    def normalize_local_data(stack_data, parent_name)
-      return nil if stack_data.nil? || stack_data.empty?
+    attr_reader :old_data, :new_data, :changes
 
-      depends = stack_data.delete('depends') || []
-      normalized_data = {
-        parent_name => stack_data.merge(
-          :loader => loader_class.for(stack_data['stack'])
-        )
-      }
+    def execute
+      set_env_variables(stack_name, current_grid)
 
-      depends.each do |stack|
-        key = "#{parent_name}-#{stack['name']}"
-        normalized_data.merge!(normalize_local_data(stack.merge('parent_name' => parent_name), key))
-      end
-      normalized_data
-    end
+      gather_data
+      process_data
+      display_report
 
-    def normalize_master_data(stack_name, raise_not_found = false)
-      begin
-        data = fetch_master_data(stack_name)
-      rescue Kontena::Errors::StandardError => ex
-        return nil if ex.status == 404 && !raise_not_found
-        raise ex
-      end
-      depends = data.delete('children') || []
+      return if dry_run?
 
-      normalized_data = { stack_name => data }
+      get_confirmation
 
-      return normalized_data if skip_dependencies?
+      run_removes
+      run_installs
+      run_upgrades
+      run_deploys if deploy?
```

Comment:
I wonder what the resulting order of parent/child stack remove/install/upgrade/deploy operations here is now, because it's slightly different than the current situation... also not entirely sure what it should be, or if there even is any easily defined correct ordering. If the parent stack services link to any of the child stack services, then the ordering will actually matter, like:

* removing a child stack + the link from the parent stack service to the child stack service => the child stack remove will fail until the parent stack is updated *and* deployed to remove the linking
* adding a child stack + linking to it from a parent service => the child stack install must run before the parent stack upgrade